Issue #484: Add basic probability aggregation capability

jessalynnsebastian opened the pull request:
## Description

This PR closes #484. It allows the user to input a list of aggregation matrices (see #482) to explicitly model a fixed reporting cycle (see #480).

## Checklist

- [x] My PR is based on a package issue and I have explicitly linked it.
- [x] I have included the target issue or issues in the PR title in the for Issue(s) *issue-numbers*: PR title
- [x] I have read the [contribution guidelines](https://github.com/epinowcast/.github/blob/main/CONTRIBUTING.md).
- [ ] I have tested my changes locally.
- [ ] I have added or updated unit tests where necessary.
- [ ] I have updated the documentation if required.
- [ ] My code follows the established coding standards.
- [ ] I have added a news item linked to this PR.
- [ ] I have reviewed CI checks for this PR and addressed them as far as I am able.

<!-- Thanks again for this PR - @epinowcast dev team -->

seabbs commented on R/model-modules.R:
```diff
@@ -247,6 +241,19 @@ enw_report <- function(non_parametric = ~0, structural = ~0, data) {
+      "The structural reporting model is in experimental development"
+    )
+    data_list$agg_probs <- 1
+    # "structural" should be a 2d (nested) list of matricess
```
I'd suggest even at this stage we do a basic check for this. we know it needs to be group by reference time long and we have that information here. we also know it needs to be a D by D matrix

seabbs commented on inst/examples/germany_reporting_cycle_model.R:
```diff
@@ -0,0 +1,124 @@
+  dplyr::mutate(confirm = new_confirm) |>
+  # Aggregate rolling sum sums over "confirm"
+  # but we want a sum over "new_confirm"
+  epinowcast:::aggregate_rolling_sum(
```
this is starting to be used a bit in places do you think we should export it?

jessalynnsebastian replied:
No strong opinions on this, I think I've only used it to create reporting cycle style data, which I don't think is that common to need to do

seabbs commented on inst/examples/germany_reporting_cycle_model.R:
```diff
@@ -0,0 +1,124 @@
+                                 1,
+                                 0)) |>
+  # Get confirm as cumulative again
+  dplyr::mutate(confirm = cumsum(confirm), .by = "reference_date")
```
this is fine but there is also a add_cumulative we could be using

seabbs commented on inst/examples/germany_reporting_cycle_model.R:
```diff
@@ -0,0 +1,124 @@
+# is reported on Wednesday.
+# Scratch in a function to make the agg_indicator for each wday
+generate_agg_indicator <- function(reference_dow_idx, reporting_dow_idx, max_delay) {
+  # Indices are as from lubridate::wday
```
This looks good to me. If it were me I would do it with dataframes though and then cast it out. Having said that I don't see a way to do that which is immediately obvious.

seabbs commented on inst/stan/epinowcast.stan:
```diff
@@ -118,6 +118,9 @@ data {
   matrix[rep_fncol, rep_rncol + 1] rep_rdesign; 
   array[2, 1] real rep_beta_sd_p;
+  // Probability aggregation indicators
+  int agg_probs;
```
we should probably assign this a prefix like repagg_probs and co to match the others?

seabbs commented on inst/stan/epinowcast.stan:
```diff
@@ -118,6 +118,9 @@ data {
   array[2, 1] real rep_beta_sd_p;
+  // Probability aggregation indicators
+  int agg_probs;
+  array[g, s] matrix agg_indicators;
```
So this may have been my confusion but s is I think the number of unique snapshots regardless of group and t is reference time. 

I would expect this to throw a bunch of out of bounds errors for that reason?

I think we have a few options. 1. Use t instead and I think this will work and 2 store this by s (and not g and t) and then in `expected_obs_from_snap` we can use that we are looping over s to pull out the right matrix. Doing that might be a little tricky as you would need to construct the write list structure in the input (or collapse the user supplied one) so perhaps 1 is the way to go

jessalynnsebastian replied:
Ah, you found this even though I forgot to tag you! Okay, got it, I misunderstood. I think using t instead makes sense.

seabbs commented on inst/stan/functions/expected_obs_from.stan:
```diff
@@ -143,7 +141,7 @@ vector expected_obs_from_snaps(int start, int end, array[] vector imp_obs,
       log_exp_obs[ssnap:esnap] = expected_obs_from_index(
         i, imp_obs, rdlurd, srdlh, refp_lh, dpmfs, ref_p, rep_h, ref_as_p, g, t,
-        l, refnp_lh, ref_np, p
+        l, refnp_lh, ref_np, p, agg_probs, agg_indicators
```
See the index lookuip on 133-136 for a sense of the information we have to play with here.

seabbs reviewed (commented):
This is looking really nice. I think I have caused some index confusion which I have tried to clarify but otherwise very close.

jessalynnsebastian commented on R/model-modules.R:
```diff
@@ -247,6 +241,28 @@ enw_report <- function(non_parametric = ~0, structural = ~0, data) {
+    cli::cli_alert_warning(
+      "The structural reporting model is in experimental development"
+    )
+    if (!(length(structural) == data$groups)
+      || !(length(structural[[1]]) == data$time)
+      || !(all(dim(structural[[1]][[1]]) == rep(data$max_delay, 2)))) {
```
Implemented a basic check here for the format of `structural` but I would have to assume there's a better way to do it

jessalynnsebastian commented on inst/stan/functions/expected_obs_from.stan:
```diff
@@ -143,7 +142,7 @@ vector expected_obs_from_snaps(int start, int end, array[] vector imp_obs,
       log_exp_obs[ssnap:esnap] = expected_obs_from_index(
         i, imp_obs, rdlurd, srdlh, refp_lh, dpmfs, ref_p, rep_h, ref_as_p, g, t,
-        l, refnp_lh, ref_np, p
+        l, refnp_lh, ref_np, p, rep_agg_probs, rep_agg_indicators[g, t][1:l, 1:l]
```
added the indexing to expected_obs_from_snaps() here, assuming I just needed to pull it "up" a level from expected_obs_from_index? difficult for me to wrap my head around what to be pulling where

seabbs replied:
This looks right to me

jessalynnsebastian commented:
@seabbs Addressed some of your comments above. When I try to compile I now have the message that `All elements in list 'rep_agg_indicators' must be numeric!` and wish I could find more documentation on passing a list of lists

seabbs commented:
>  your comments above. When I try to compile I now have the message

When you compile or try and fit? I wouldn't have thought the input data would have an interaction with compiling but just when you try and fit the data?

seabbs commented:
I've just sight read this so far so will install and try it out

jessalynnsebastian commented:
Yep, my bad - not when compiling, when fitting a model.

jessalynnsebastian commented on inst/examples/germany_reporting_cycle_model.R:
```diff
@@ -105,13 +105,12 @@ generate_agg_indicator <- function(reference_dow_idx, reporting_dow_idx, max_del
-agg_indicators[[1]] <- lapply(seq_len(length(ref_t)), function(t) {
-  return(ref_day_matrix_list[[ref_t_wdays[t]]])
-})
+agg_indicators <- array(dim = c(1, length(ref_t), 35, 35))
+for (t in seq_along(ref_t)) {
+  agg_indicators[1, t, , ] <- ref_day_matrix_list[[ref_t_wdays[t]]]
+}
```
@seabbs Formatting agg_indicators like this dealt with that specific error about numeric lists, but now I can't get Stan to initialize when trying to fit a basic model. I'll keep messing with it.

seabbs commented:
I'll try and give this a try today and see if I can see anything

seabbs commented:
So I think the remaining issue here is probably in input preparation as the rest of the changes look good to me. I am just writing another version of that preprocessing to see if I can find an issue but so far I haven't seen one (boy this input format is hard to grasp right now). 

The reason I think its this is the error is it sees a -Inf (i.e a 0 prob) that makes me think there is mismatch between the .observed mask and the accumulation matrix. This also means this really can't ever be used where you tried and observe the input so for anything user facing we would need to think about that (as it is super tricky as above)

seabbs commented:
Once this is passing checks will merge into the dev version
